tests/_backends/test_anyio.py

```diff
@@ -0,0 +1,96 @@
+    backend = httpcore.AnyIOBackend()
+    stream = await backend.connect_tcp(
+        httpbin.host, httpbin.port, timeout=CONNECT_TIMEOUT
+    )
+    await stream.aclose()
+
+
+@pytest.mark.anyio
+async def test_write_without_tls(httpbin):
+    backend = httpcore.AnyIOBackend()
+    stream = await backend.connect_tcp(
+        httpbin.host, httpbin.port, timeout=CONNECT_TIMEOUT
+    )
+    async with stream:
+        http_request = [b"GET / HTTP/1.1\r\n", b"\r\n"]
+        for chunk in http_request:
+            await stream.write(chunk, timeout=WRITE_TIMEOUT)
+
+
+@pytest.mark.anyio
+async def test_read_without_tls(httpbin):
+    backend = httpcore.AnyIOBackend()
+    stream = await backend.connect_tcp(
+        httpbin.host, httpbin.port, timeout=CONNECT_TIMEOUT
+    )
+    http_request = [b"GET / HTTP/1.1\r\n", b"\r\n"]
+    async with stream:
+        for chunk in http_request:
+            await stream.write(chunk, timeout=WRITE_TIMEOUT)
+        await stream.read(1024, timeout=READ_TIMEOUT)
+
+
+@pytest.mark.anyio
+async def test_connect_with_tls(httpbin_secure):
+    backend = httpcore.AnyIOBackend()
+    ssl_context = ssl.create_default_context()
+    ssl_context.load_verify_locations(certs.where())
+    stream = await backend.connect_tcp(
+        httpbin_secure.host, httpbin_secure.port, timeout=CONNECT_TIMEOUT
+    )
+    async with stream:
+        tls_stream = await stream.start_tls(
+            ssl_context=ssl_context, timeout=CONNECT_TIMEOUT
+        )
+        await tls_stream.aclose()
+
+
+@pytest.mark.anyio
+async def test_write_with_tls(httpbin_secure):
+    backend = httpcore.AnyIOBackend()
+    ssl_context = ssl.create_default_context()
+    ssl_context.load_verify_locations(certs.where())
+    stream = await backend.connect_tcp(
+        httpbin_secure.host, httpbin_secure.port, timeout=CONNECT_TIMEOUT
+    )
+    async with stream:
+        tls_stream = await stream.start_tls(
+            ssl_context=ssl_context, timeout=CONNECT_TIMEOUT
+        )
+        async with tls_stream:
+            http_request = [b"GET / HTTP/1.1\r\n", b"\r\n"]
+            for chunk in http_request:
+                await tls_stream.write(chunk, timeout=WRITE_TIMEOUT)
+
+
+@pytest.mark.anyio
+async def test_read_with_tls(httpbin_secure):
+    backend = httpcore.AnyIOBackend()
+    ssl_context = ssl.create_default_context()
+    ssl_context.load_verify_locations(certs.where())
+    stream = await backend.connect_tcp(
+        httpbin_secure.host, httpbin_secure.port, timeout=CONNECT_TIMEOUT
+    )
+    async with stream:
+        tls_stream = await stream.start_tls(
+            ssl_context=ssl_context, timeout=CONNECT_TIMEOUT
+        )
+        async with tls_stream:
+            http_request = [b"GET / HTTP/1.1\r\n", b"\r\n"]
+            for chunk in http_request:
+                await tls_stream.write(chunk, timeout=WRITE_TIMEOUT)
+            await tls_stream.read(1024, timeout=READ_TIMEOUT)
```

Comment:
This set of tests are looking good. I wonder if we could drop `httpbin` and instead test against plain old socket code? We'll probably need to do that in order to be able to add testing for the TLS-in-TLS case.

Reply:
For simplicity, we can use httpbin for these tests and write a fixture that implements simple tls-in-tls tunneling.

Reply:
We can use something like this to remove httpbin and test tls-in-tls.
```python
def serve(sock):
    with client_socket as sock.accept()[0]:
        while True:
            data = client_socket.recv(1024)
            if not data:
                client_socket.close()
                break
            client_socket.send_all(data)

@pytest.fixture
def tls_server():
    with socket.socket() as socket:
        sock.bind((IP, PORT))
        sock.listen()
        with context.wrap_socket(sock) as tls_sock:
            Thread(target=serve, args=[tls_sock]).start()
            yield tls_sock

tls_server_copy = tls_server

@pytest.fixture
def tls_in_tls_tunnel(tls_server, tls_server_copy):
    Thread(target=tunnel, args=[tls_server, tls_server_copy]).start()
    yield tls_server
```

Reply:
We can use something like this to remove httpbin and test tls-in-tls.
```python
def serve(sock):
    with client_socket as sock.accept()[0]:
        while True:
            data = client_socket.recv(1024)
            if not data:
                client_socket.close()
                break
            client_socket.send_all(data)

@pytest.fixture
def tls_server_socket():
    with socket.socket() as socket:
        sock.bind((IP, PORT))
        sock.listen()
        with context.wrap_socket(sock) as tls_sock:
            yield tls_sock

@pytest.fixture
def tls_server(tls_server_socket):
    Thread(server, tls_server_socket).start()
    yield tls_server_socket.getsocketname()  # ip, port tuple

tls_server_socket_copy = tls_server_socket

@pytest.fixture
def tls_in_tls_tunnel(tls_server_socket, tls_server_socket_copy):
    Thread(target=tunnel, args=[tls_server_socket, tls_server_socket_copy]).start()
    yield tls_server
```

Reply:
But, should we use threads to solve this? It appears to be somewhat complicated.

Reply:
I implemented it using socket logic. if it appears too complicated, we can revert it, but I believe it looks good.